src/status_im/chat/commands/core.cljs

```diff
@@ -0,0 +1,139 @@
+(ns status-im.chat.commands.core
+  (:require-macros [status-im.utils.views :refer [defview letsubs]])
+  (:require [re-frame.core :as re-frame]
+            [status-im.ui.components.react :as react]
+            [status-im.ui.components.icons.vector-icons :as vector-icons]
+            [status-im.ui.components.colors :as colors]
+            [status-im.i18n :as i18n]
+            [status-im.chat.styles.message.message :as style]))
+
+(defprotocol Command
+  "Protocol for defining command message behaviour"
+  (id [this] "Identifier of the command, used to look-up command display name as well")
+  (scope [this]
+    "Scope of the command, defined as set of values representing contexts
+     where in which command is available, together with `id` it forms unique
+     identifier for each command.
+     Available values for the set are: 
+     `id-of-the-any-chat` - command if available only for the specified chat
+     `:personal-chats` - command is available for any personal 1-1 chat
+     `:group-chats` - command is available for any group chat
+     `:public-chats` - command is available for any public chat
+     `:requested` - command is available only when there is an outstanding request")
+  (parameters [this]
+    "Ordered sequence of command parameter templates, where each parameter 
+     is defined as map consisting of mandatory `:id`, `:title` and `:type` keys, 
+     and optional `:suggestions` field.
+     When used, `:suggestions` containes reference to any generic helper component
+     rendering suggestions for the argument (input code will handle when and where
+     to render it)")
+  (validate [this parameters cofx]
+    "Function validating the parameters once command is send. Takes parameters map
+    and `cofx` map as argument, returns either `nil` meaning that no errors were
+    found and command send workflow can proceed, or sequence of errors to display")
+  (on-send [this message-id parameters cofx]
+    "Function which can provide any extra effects to be produced in addition to 
+    normal message effects which happen whenever message is sent")
+  (short-preview [this command-message]
```

Comment:
Are both `preview` and `short-preview` required for any command? Could this be defined in a separate protocol?

Reply:
Basically yes, every command message needs to define how `short-preview` and `preview` are rendered. 
The way to generalise such behaviour would be to have some generic function, which will just take the command name and generate nicely formatted short-preview/preview when no special handling (like generating "live" preview with tx confirmations for `/send` command) is necessary.

Reply:
Previews are for messages right? Would it make sense to have message free commands?

Reply:
Correct, command-messages are just messages with special formatting and semantic attached to them (now when I think about it, we should add `on-receive` method to the protocol, that way we could express functionality of the `/request` commands super nicely, without any special branching in general code).
Message free commands are quite different semantically (for example there is no reason why message free commands should be tied to particular chat)  and I don't think it makes sense to fit those two into the same box.

Reply:
I still think we should decouple messages from commands. There is at least one use case for a message sent without the command - chat message after wallet send.

Reply:
On the related topic, I think we should stay away from message-free commands (e.g. like `@browse`). Agree that there is nothing to tie the command to any chat in that case. The same functionality can be provided by a regular dapp and especially so if we provide some API to access contacts

Reply:
On a less related topic, can we take this opportunity to rename `preview` and `short-preview` to something more appropriate? The current names make no sense - a preview is something you see before, not after the message is sent.